dolittle/Studio
# Introducing navigation across portal and microservices

![2020-12-16_19-46-53 (1)](https://user-images.githubusercontent.com/134365/102394040-76e36800-3fd9-11eb-9c3a-40cbb508a415.gif)


This introduces the concept of having navigation that is consistent from the top level window to the inner iframe documents. This has been built with our convention of `/<microservice>` to `/_/<microservice>` for iframes in mind.

Changes that are introduced:
* Bootstrapper now includes a router - one does not need to set this up
* Switch is not used within Microservices for routing - as we need more control
* In the Portal we now have a new construct called `<CompositionRoute/>` that renders the correct iframe and deals with reloads and setting correct `src` for the iframe
* `<MicroserviceRoute/>` is used within a Microservice to tell what sub routes it is listening to. It does not need a full path as this is derived from the `MicroserviceContext``. It is optional whether or not this is used - depends on the microservice.
* A React Context is introduced; `MicroserviceContext` that holds the configuration given in the Bootstrapper and makes it available from anywhere - this is used to derive the prefix for any routes within a Microservice.
* `<Routing/>` is used within a Microservice and all `<MicroserviceRoute/>` instances lives within it. This governs the actual pushing of state to the iframe windows history.
* `NavigationItem` introduces a new optional property called `path`. Once this is set, navigation will work with the new `INavigator`. The path one puts into this is the top level path, not the one with the `/_` prefix - for anything within **Applications** this would be `/applications/something`.
* Message `NavigatedTo` gets published - putting in a dependency to `IMessenger` in a ViewModel lets you listen to this. Useful for those scenarios where the navigation is navigating data and not using the `<MicroserviceRoute/>`.

## How to test

1. Run environment according to [this](https://github.com/dolittle/Studio/blob/development/Documentation/getting-started.md)
2. Start portal Web as explained [here](https://github.com/dolittle/Studio/blob/development/Documentation/getting-started.md)
3. Start applications Web as explained [here](https://github.com/dolittle/Studio/blob/development/Documentation/getting-started.md)
4. Optionally start applications Backend as explained [here](https://github.com/dolittle/Studio/blob/development/Documentation/getting-started.md)
5. Navigate to [http://localhost:9000/applications](http://localhost:9000/applications) - you should see "This is the roooot" in the iframe content
6. Change the URL to [http://localhost:9000/applications/something](http://localhost:9000/applications/something) - you should see "Something" in the iframe content. (this is effectively same as refresh)

## Missing things

The biggest missing feature right now is probably that history navigation does not work (back and forward). We will have to revisit this.

The following tasks are known that is left overs / technical debt:
- Use navigator in the top level menu
- Navigation structure needs to set the correct item when loading based on current location path
- Fix back and forward for navigation
- Add a way to get a callback when navigation has occurred
- Add data and title to the navigator when navigating
- Add lifecycle events into the navigator (Before, After)

These are registered [here](https://app.asana.com/0/1190518762954630/1199621706420926)



## Challenge

The name `<MicroserviceRoute/>` feels wrong. We have a top level `<CompositionRoute/>` and I'd love a better name for whatever sits inside it. See this as a hierarchy.
